### PR TITLE
Completer refactor.

### PR DESCRIPTION
**Pull request by afshin**
This PR addresses the issues raised in https://github.com/jupyter/jupyterlab/issues/847.

## Bugfixes
- [x] When completer height exactly equals max height, there is a horizontal scrollbar.
- [x] Subsequent queries after the inital open of the completer are not working on multi-line code cells.
- [x] Remove uses of the `dataset` property of nodes and instead switch to `(get|set|remove)Attribute`, cf. https://github.com/phosphorjs/phosphor/pull/137.

## Maintenance
- [x] Make the completer a top-level standalone widget since it is used in both notebooks and consoles.
- [x] Refactor the completer widget and rename it from `Completion...` to `Completer...` in order to semantically align it with how we name other widgets and how we talk about functionality. In particular, places where a *completion* actually occurs are still called `completion` in the variable nomenclature, but most other places follow the naming of other "actor" classes, *e.g.*, `docmanager`, `editor`, `inspector`, *etc.*

## Behaviors

- [x] Pressing tab initiates completion for the currently typed line and displays the possible completions **with the caveat that if there is *one and only one* match**, the completer automatically fills in the text (*e.g.*, typing `pri<tab>` will result in the completer filling out the value `print`).
- [x] On initial start of a new completer session (*i.e.*, a newly invoked completion request), the common subset of possible values should be pre-selected.
- [x] Up/down arrows can be used to navigate the possible completions.
- [x] Return or a single mouse click can be used to select the currently selected completion.
- [x] If a user continues to type code while the completer is open, the typed text will be appended to the code cell and the list of completions will be filtered further.
- [x] The tab completion text is aligned with the code in the cell
- [x] Pressing esc while the completer is open should close it and focus the code cell.
- [x] The completer privileges appearing beneath the text when possible. It will automatically switch to appearing above if that is a better option (see screen capture below).

## Screenshots
Here is a screenshot showing some of these behaviors:

completer screenshot

Here's a small demo:

completer

Here is a demo of the scrolling behavior:

completer

### CODE REVIEW

**Comment by afshin** [comment]
@blink1073 @Carreau @ellisonbg 

This is ready for review.

**Comment by blink1073** [comment]
👍 overall, will let others comment on the behavior.

**Comment by Carreau** [comment]
Assigning myself to actually try it in more depth..

**Comment by afshin** [comment]
Adding "work in progress" label again to add a feature that privileges rendering the completer *below* unless there is not enough space.

File: src/completer/index.css [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  font-size: var(--jp-private-completer-code-font-size);
+  height: var(--jp-private-completer-item-height);
+  padding: 0;
+  margin: 0;
+}
+
+
+.jp-Completer-item mark {
+  font-weight: bold;
+  text-decoration: underline;
```
Comment (ellisonbg) [inline src/completer/index.css]: I think this makes is overly distracting, and with the completer UI aligned with the stuff the user typed, I don't think it is needed

File: src/completer/index.css [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+.jp-Completer-item {
+  margin: 0;
+  min-width: 150px;
+  padding: 0 2px;
+}
+
+
+.jp-Completer-item:hover, .jp-Completer-item.jp-mod-active {
+  background: #BFBFBF;
```
Comment (ellisonbg) [inline src/completer/index.css]: `var(--jp-layout-color2)`

File: src/completer/index.css [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+
+:root {
+  --jp-private-completer-code-font-size: 14px;
+  --jp-private-completer-item-height: 24px;
+}
+
+
+.jp-Completer {
+  background: white;
```
Comment (ellisonbg) [inline src/completer/index.css]: `var(--jp-layout-color1)` so dark themes can style this

File: src/completer/index.css [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  border: var(--jp-border-width) solid var(--jp-border-color1);
```
Comment (ellisonbg) [inline src/completer/index.css]: Maybe try `color2` for the border

File: src/completer/index.css [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  min-height: calc(var(--jp-private-completer-item-height) + (2 * var(--jp-border-width)));
+  position: absolute;
+  z-index: 10001;
+}
+
+
+.jp-Completer.jp-mod-outofview {
+  display: none;
+}
+
```
Comment (ellisonbg) [inline src/completer/index.css]: Let's also add a subtle drop shadow similar to how menus have...

**Pull request merged**